### PR TITLE
CI: build-once + sharded parallel test matrix (with parallel simulator execution)

### PR DESCRIPTION
**Pull request by ericmasiello**
## Summary

Follow-up to #49 (implemented via PR #71). Restructures the "Unit + UI tests" job (~12m28-58s, of which UI tests alone accounted for ~518s/~69%) into a build-once + sharded matrix so PR checks aren't serially bottlenecked on one VM.

- `build-for-testing` builds once (`xcodebuild build-for-testing`) and uploads the built products + generated `.xctestrun` as an artifact (tarred first — `actions/upload-artifact`'s own zip/unzip roundtrip is a known footgun for symlinks inside `.app`/`.xctest` bundles on macOS).
- `test` runs as a 4-way matrix, each downloading that artifact and running `xcodebuild test-without-building -only-testing:...` for its assigned classes:
  - `unit` → `SubTimerTests`
  - `ui-player-launch` → `PlayerComponentsUITests` + `SubTimerUITests` + `SubTimerUITestsLaunchTests`
  - `ui-settings` → `SettingsViewUITests`
  - `ui-timer` → `TimerViewUITests`
- Same pinned `macos-26` / Xcode `26.5` / simulator `OS=26.5` versions from #49 (ADR-0005) — unchanged.

## Parallel simulator execution: deferred (see ADR-0008)

The issue also asked for `-parallel-testing-enabled YES -maximum-concurrent-test-simulator-destinations N` within each UI shard. Two live PR runs proved this **reliably hangs** on this exact `macos-26`/Xcode `26.5` combination: `xcodebuild` clones the destination simulator, and DTServiceHub fails to resume the cloned process (`Failed to send signal 19 to process ...`), which `xcodebuild` never recovers from. Pre-booting and pinning to the base device's UDID before letting it get cloned didn't help — reproduced identically on a local Xcode 26.6 Mac too. This is a known, currently-unresolved Xcode 26.x issue on GitHub-hosted macOS runners (`actions/runner-images#13264` and linked issues) — see **ADR-0008** for full evidence and rationale.

Each shard now runs its own classes serially within itself instead. The underlying goal — PR checks not bottlenecked on one VM — still holds via job-level concurrency (4 shards on separate runners), just without the extra within-shard squeeze that clone parallelism would have added.

## Real measured results (2 consecutive live PR runs, no new flakiness)

| Job | Run 1 | Run 2 |
|---|---|---|
| SwiftLint | 15s | 15s |
| Build for testing | 1m1s | 1m53s |
| Test / unit | 5m30s | 3m55s |
| Test / ui-player-launch | 9m17s | 9m45s |
| Test / ui-settings | 10m0s | 9m49s |
| Test / ui-timer | 9m0s (after 1 flake, see below) | 6m13s |
| **Total PR-blocking wall time** | **~11m3s** | **~11m46s** |

Baseline on `main` (3 prior PRs): ~12m28–58s serial.

One `TimerViewUITests.testTimerControlsBehavior` failure occurred on the first attempt of run 1 ("Button state should change after tap") and passed clean on retry with the identical artifact/shard — a pre-existing UI-timing flake, not something introduced by this restructuring. Both full 4-shard reruns since then have been clean.

**Honest take on wall time**: this is a modest win (~1min), not the dramatic one the ticket was aiming for. Each shard now pays its own fixed overhead (Xcode select, CoreSimulator warm-up, artifact download, simulator boot) redundantly 4x instead of once, and the piece that was meant to offset that — clone-based parallelism shrinking each shard's own test time — is exactly the piece blocked by the Xcode bug above. What this change delivers regardless of wall time: per-shard failure isolation (exact failing shard + its own `.xcresult`, instead of one monolithic log), no full rebuild per test run, and the shard topology #73 needs to add label-gated perf-test tiering.

Closes #72

## Follow-up: label-gated perf test tiering (#73)

On top of the shard topology above, `ui-player-launch` now excludes the 3
XCTest `measure{}` performance tests
(`PlayerComponentsUITests.testPlayerRowsRenderPerformance`,
`PlayerComponentsUITests.testScrollingPerformance`,
`SubTimerUITests.testLaunchPerformance`) by default via `-skip-testing:` -
they're inherently slow by design and signal "did it get slower," not "did
it break." Applying the `test-ui-performance` label to this PR re-triggers
the check (the `pull_request` trigger now includes `labeled`) and runs the
full suite, including those 3 tests. Removing the label reverts subsequent
triggers to the default (fast) tier. Documented in README's CI section.

Closes #73

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -60,10 +68,148 @@ jobs:
...
+      - name: Build for testing (SubTimerTests + SubTimerUITests)
         run: |
-          xcodebuild test \
+          xcodebuild build-for-testing \
             -project SubTimer.xcodeproj \
             -scheme SubTimer \
             -testPlan SubTimerCI \
-            -destination "${{ env.SIMULATOR_DESTINATION }}"
+            -destination "${{ env.SIMULATOR_DESTINATION }}" \
+            -derivedDataPath "${{ env.DERIVED_DATA_PATH }}"
```
[the remark this answers is not in the document]
Comment (ericmasiello) [reply .github/workflows/ci.yml:90]: Good catch — this was already fixed on this branch in commit 45c4afe ("ci: fix DerivedData cache path mismatch; correct stale ADR-0005 text"), which changed the cache step's `path:` to `${{ env.DERIVED_DATA_PATH }}` to match `-derivedDataPath`. Resolving.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -60,10 +68,148 @@ jobs:
...
+
+      - name: Locate the generated .xctestrun
+        id: locate-xctestrun
+        run: |
+          xctestrun_path=$(find "${{ env.DERIVED_DATA_PATH }}/Build/Products" -maxdepth 1 -name '*.xctestrun')
+          if [ -z "$xctestrun_path" ]; then
+            echo "::error::No .xctestrun file found under Build/Products" >&2
+            exit 1
+          fi
+          echo "name=$(basename "$xctestrun_path")" >> "$GITHUB_OUTPUT"
```
[the remark this answers is not in the document]
Comment (ericmasiello) [reply .github/workflows/ci.yml]: Fixed in 12e2011 — the stale .xctestrun is now deleted before the build runs (the cache carrying over an old one is the realistic way this could have ever matched >1 file), and the locate step now fails loudly if it ever finds anything other than exactly one match instead of silently basename-ing whatever `find` returned.
